### ISSUE TITLE
Enum schema (list of values) not written in discovery file anymore

### ISSUE THREAD

**Issue by clementdenis** [cloudendpoints/endpoints-java#49]
Since beta.10, the enum values are not written to the discovery files anymore.
This is a problem, as enum values provided in the discovery can be used in frontend code to define possible values in select fields.

Here is an example:
- The [Api class](https://gist.github.com/clementdenis/305f593028d60b1295ca5c099fa38fd6#file-specialfeaturesapi-java) and the [enum used as param](https://gist.github.com/clementdenis/305f593028d60b1295ca5c099fa38fd6#file-testenum-java)
- Discovery generated by beta.9 => contains the [enum values](https://gist.github.com/clementdenis/305f593028d60b1295ca5c099fa38fd6#file-discovery-2-0-0-beta-9-json-L118)
- Discovery generated by 2.0.0 final => enum values not present in [schemas](https://gist.github.com/clementdenis/305f593028d60b1295ca5c099fa38fd6#file-discovery-2-0-0-json-L113)

**Comment by tangiel** [comment]
This is interesting, because we specifically have a test case for this, and it's passing. Can you include `TestResource`, even though it's probably trivial?

**Comment by clementdenis** [comment]
It is indeed trivial, I [added it anyway](https://gist.github.com/clementdenis/305f593028d60b1295ca5c099fa38fd6#file-testresource-java).
I made some more tests, and the issue happens if the TestResource (including TestEnum as a field type) is used ONLY as output resource. If I use TestResource as an input resource, the enum values are written again.
This is indeed a corner case, as an API usually has both read and write methods for a resource.

**Comment by tangiel** [comment]
The test case in the repository right now actually only has the resource in the return value:

```java
@Api(name = "enum", version = "v1")
public class EnumEndpoint {
  @ApiMethod(name = "create", path = "{value}")
  public EnumValue create(@Named("value") TestEnum value) {
    return null;
  }

  private static class EnumValue {
    public TestEnum value;
  }
}
```

**Comment by clementdenis** [comment]
I could finally pinpoint the cause of the issue:
- I have multiple APIs running in the same app
- TestResource is reused in more than one of these APIs as a resource
- Only the FIRST API using the resource with an enum that has its discovery generated gets the TestEnum schema, the others don't

Once you know that, the problem becomes obvious: the [types fields in SchemaRepository](https://github.com/cloudendpoints/endpoints-java/blob/master/endpoints-framework/src/main/java/com/google/api/server/spi/config/model/SchemaRepository.java#L47) is using an ApiSerializationConfig as the key, which is the same instance for all APIs, so schemas are cached between APIs.
The problem is that [when using the cached value](https://github.com/cloudendpoints/endpoints-java/blob/master/endpoints-framework/src/main/java/com/google/api/server/spi/config/model/SchemaRepository.java#L119), only the resource schema is added to the API schemas, not its field's schemas.
So the problem is not limited to enum values, but actually to any resource field's schema that would be reused in multiple APIs.

**Comment by tangiel** [comment]
Got it, great find! I have a fix in #53.

**Comment by tangiel** [comment]
This is now fixed.
